src/gatsby/models/authors/parse-link.js

```diff
@@ -0,0 +1,94 @@
+/*
+   Processor builder: makes a function that turns a string into processed Link metadata
+
+   Takes "Mod functions" as parameters, returns a function that starts with {} and pipes
+   that through each function, then returns the result.
+
+   signature: ({ scheme: string, host: string, pathname: string}, input)
+   example:      ^ https://      ^ mysite.com  ^ /rest/of/the/url ^ an object
+
+   The second parameter will be the current state of the input, and usually will
+   have to be spread into your return result.
+ */
+function processor(...mods) {
+  return function process(groups) {
+    return mods.reduce(
+      (currentResult, _, i) => mods[i](groups, currentResult),
+      {}
+    )
+  }
+}
+
+// Processor helpers
+
+function trimSlashes(input) {
+  return /^\/?(.*?)\/?$/.exec(input)[1]
+}
+
+const asSite = site => (groups, input) => ({
+  ...input,
+  site
+})
+
+const pathnameAsUsername = ({ pathname }, input) => ({
+  ...input,
+  username: trimSlashes(pathname)
+})
+
+const urlHTTPS = ({ host, pathname }, input) => ({
+  ...input,
+  url: 'https://' + host + pathname
+})
+
+const urlDefault = ({ scheme, host, pathname }, input) => ({
+  ...input,
+  url: (scheme || 'https://') + host + pathname
+})
+
+// Building processors for recognized sites
+
+const processors = {
+  'twitter.com': processor(asSite('twitter'), urlHTTPS, pathnameAsUsername),
+  'github.com': processor(asSite('github'), urlHTTPS, pathnameAsUsername),
+  'linkedin.com': processor(
+    // Handle LinkedIn as a special case
+    ({ pathname }) => ({
+      site: 'linkedin',
+      username: /^\/in\/(.*)/.exec(pathname)[1]
+    }),
+    urlHTTPS
+  )
+}
+
+const defaultProcessor = processor(asSite(null), urlDefault)
+
+function WrongTypeError(type) {
+  return new Error(`A ${type} cannot be used as input for parseLink!`)
+}
+
+function parseLink(input) {
+  switch (typeof input) {
+    // Handle shorthand string links
+    case 'string':
+      const {
+        groups
+      } = /^(?<scheme>.*?\/\/)?(?:www\.)?(?<host>[^\/]*)(?<pathname>.*?)\/?$/.exec(
```

Comment:
ok, makes sense regarding using regex.

just a few thing to check

do we actually need `\/?` at the end - I think `.*` captures everything enyway.

can we normalize pathname right here (like we essentially do with `www.`) - trim `/`. instead of using additional helper?

Reply:
`\/?` is specifically not captured, and removes trailing slashes from the URL.

```
can we normalize pathname right here (like we essentially do with www.) - trim /. instead of using additional helper?
```

That's exactly what I'm doing here. `www.` is trimmed to make `website.com` and `www.website.com` equivalent, and this regex also handles trailing slashes. I did my best to frontload all the work I could into this regex, down to the named capture groups because positional ones break with how much it's doing.

Reply:
you should be able to simplify trimSlashes into trimLeadingSlash if the trailing is trimmed here?

Reply:
Ah, yes, I can do that. Coming up in the next push!

Reply:
Done! It's embedded now and simplified to only work on leading slashes.